gitenberg/dbUpdates/updateDb.py

```diff
@@ -0,0 +1,123 @@
+import psycopg2
+import yaml
+import os
+import six # check file types
+from gitenberg import book as Book
+from gitenberg.metadata.pg_rdf import pg_rdf_to_yaml
+from gitenberg.util.catalog import BookMetadata
+from gitenberg.util import util
+
+# for parsing yaml tags
+def default_ctor(loader, tag_suffix, node):
+    return tag_suffix + ' ' + node.value
+yaml.add_multi_constructor('!lcc', default_ctor)
+yaml.add_multi_constructor('!lcsh', default_ctor)
+
+
+NUM_BOOK_COLS = 11 # database table constant
+NUM_AUTHOR_COLS = 6
+conn = psycopg2.connect("dbname=stuff user=marc")
+cur = conn.cursor()
+
+def updateDB(repo_name, sqlCur, sqlConn):
+    book = Book.Book(None, repo_name=repo_name)
+    # clone book if needed
+    if not book.local_repo:
+        book.clone_from_github()
+        
+    # fast forward only on server, should be fine
+    book.local_repo.git.remote("origin").pull() 
+    with util.cd(book.local_repo.repo_path): # path to working dir of repo
+        rdfFile = "pg" + book.book_id + ".rdf"
+        if os.path.isfile("metadata.yaml"):
+            with open("metadata.yaml") as f:
+                data = yaml.load(f)
+                updateBook(sqlCur, data, repo_name)
+                updateAuthor(sqlCur, data)
+                # TODO gather cover metadata
+                sqlConn.commit()
+        elif os.path.isfile(rdfFile):
+            createMetaFile(book)
+
+def updateBook(sqlCur, data, repo_name):
+    sqlExistsBook = "SELECT count(*) FROM Book WHERE book_id = %s"
+    sqlInsertBook = "INSERT INTO book VALUES (" + \
+                    "%s, "*(NUM_BOOK_COLS-1) + "%s)"
+    sqlUpdate = """UPDATE book SET
+        (repo_name, title, language, description, author, gutenberg_type,
+         gutenberg_bookshelf, subjects, full_text, num_downloads) = (""" + \
+             "%s, "*(NUM_BOOK_COLS-2) + "%s)" + \
+             "WHERE book_id=%s"
+
+    # update/insert book
+    sqlCur.execute(sqlExistsBook, (data['identifiers']['gutenberg'],))
+    count = sqlCur.fetchone()
+    bookRecord = makeBookRecordFromRepo(repo_name, data)
+    if count[0] == 1:
+        sqlCur.execute(sqlUpdate, bookRecord[1:] + [bookRecord[0]])
+    elif count[0] == 0:
+        sqlCur.execute(sqlInsertBook, bookRecord)
+
+def updateAuthor(sqlCur, data):
+    sqlExistsAuthor = "SELECT count(*) FROM Author WHERE author_id = %s"
+    sqlInsertAuthor = "INSERT INTO author VALUES (" + \
+                      "%s, "*(NUM_AUTHOR_COLS-1) + "%s)"
+    
+    authorRecord = makeAuthorRecordFromRepo(data)
+    sqlCur.execute(sqlExistsAuthor, (authorRecord[0],)) # 0 holds id
+    count = sqlCur.fetchone()
+    if count[0] == 0: # insert author
+        sqlCur.execute(sqlInsertAuthor, authorRecord)
+    # TODO update author
+
+def createMetaFile(book):
```

Comment:
the workflow for metadata file creation was not well documented, mostly because the strategy was not clear. Have you thought out a workflow for creating the metadata files? Note that "enrich" is an expensive operation because it queries wikipedia data. That's why it's an option elsewhere- you want do it infrequently.

Reply:
The idea is that this will happen only once per book, when there is a git repo which has only rdf metadata and not yaml. This function is intended to be called via a queuing service, so it will not be called too often, all in all this will be used infrequently, but we do want to ensure that there is a description for every book, as much as possible.

I can add more documentation for the workflow, but it mainly uses existing functionality, it seemed fairly straightforward.

Reply:
well the existing workflow was ad hoc. The idea was to create the metadata file it during the process of creating the repo. rsync the pg files locally, add the metadata, then push to github. The problem was what to do when updating/enhancing the repo. You want to be sparing in the use of the github api, so instead we'd git pull, manipulate, then git push. Where it gets complicated is when you have things happening from different places- command line, website, etc. So thinking out and documenting the workflow is very important here. There was no workflow for a metadata-on-commit process, though I assume that will be needed.

Reply:
So would you like a separate file with documentation for that? The full workflow uses the other process which handles git hooks and the queuing service, so it shouldn't be only part of this code file I think.